micro_http/src/connection.rs

```diff
@@ -0,0 +1,714 @@
+            }
+            self.read_cursor = 0;
+
+            return Ok(false);
+        }
+
+        // Append only the remaining necessary bytes to the body of the request.
+        self.body_vec.extend_from_slice(
+            &self.buffer
+                [*line_start_index..(*line_start_index + self.body_bytes_to_be_read as usize)],
+        );
+        *line_start_index += self.body_bytes_to_be_read as usize;
+        self.body_bytes_to_be_read = 0;
+
+        let request = self.pending_request.as_mut().unwrap();
+        // If there are no more bytes to be read for this request.
+        // Assign the body of the request.
+        let placeholder: Vec<_> = self
+            .body_vec
+            .drain(..request.headers.content_length() as usize)
+            .collect();
+        request.body = Some(Body::new(placeholder));
+
+        // If we read more bytes than we should have into the body of the request.
+        if !self.body_vec.is_empty() {
+            return Err(ConnectionError::StreamError);
+        }
+
+        self.state = ConnectionState::RequestReady;
+        Ok(true)
+    }
+
+    /// Tries to write responses to the provided stream.
+    /// Meant to be used only with non-blocking streams and an `EPOLL` structure.
+    /// Should be called whenever an `EPOLLOUT` event is signaled.
+    pub fn try_write(&mut self) -> Result<(), ConnectionError> {
+        if self.response_buffer.is_none() {
+            if let Some(response) = self.response_queue.pop_front() {
+                let mut response_buffer_vec: Vec<u8> = Vec::new();
+                response
+                    .write_all(&mut response_buffer_vec)
+                    .map_err(|_| ConnectionError::StreamError)?;
+                self.response_buffer = Some(response_buffer_vec);
+            } else {
+                return Ok(());
+            }
+        }
+
+        let mut response_fully_written = false;
+
+        if let Some(response_buffer_vec) = self.response_buffer.as_mut() {
+            let bytes_to_be_written = response_buffer_vec.len();
+            match self.stream.write(response_buffer_vec.as_slice()) {
+                Ok(0) => {
+                    return Err(ConnectionError::ClosedConnection);
+                }
+                Ok(bytes_written) => {
+                    if bytes_written != bytes_to_be_written {
+                        response_buffer_vec.drain(..bytes_written);
+                    } else {
+                        response_fully_written = true;
+                    }
+                }
+                Err(_) => {
+                    return Err(ConnectionError::StreamError);
+                }
+            }
+        }
+
+        if response_fully_written {
+            self.response_buffer.take();
+        }
+
+        Ok(())
+    }
+
+    /// Send a response back to the source of a request.
+    pub fn send_response(&mut self, response: Response) {
+        self.response_queue.push_back(response);
```

Comment:
Is this really sending the response? My understanding is that it just queues the response and then the user has to call `try_write()`?

Reply:
Indeed, this should have a name like `enqueue_response()`.